pkg/tftarget/tftarget_test.go

```diff
@@ -0,0 +1,270 @@
+}
+
+func stringToInterface(s []string) []interface{} {
+	iface := make([]interface{}, len(s))
+	for i := range s {
+		iface[i] = s[i]
+	}
+	return iface
+}
+
+// includeAddress populates the includeAddress field inside of the match block
+func includeAddress(val ...string) func(map[string]interface{}) {
+	return func(matchBlock map[string]interface{}) {
+		matchBlock["addresses"] = stringToInterface(val)
+	}
+}
+
+// excludeAddress populates the excludeAddress field inside of the match block
+func excludeAddress(val ...string) func(map[string]interface{}) {
+	return func(matchBlock map[string]interface{}) {
+		matchBlock["excludedAddresses"] = stringToInterface(val)
+	}
+}
+
+// reviewTestData is the base test data which will be manifested into a
+// raw JSON.
+type reviewTestData struct {
+	name                string
+	match               map[string]interface{}
+	address             string
+	wantMatch           bool
+	wantConstraintError bool
+	providerName        string
+	removeProviderBlock bool
+}
+
+func (td *reviewTestData) jsonAssetTestcase() *targettesting.ReviewTestcase {
+	tc := &targettesting.ReviewTestcase{
+		Name:                td.name,
+		Match:               td.match,
+		WantMatch:           td.wantMatch,
+		WantConstraintError: td.wantConstraintError,
+	}
+	if td.match != nil {
+		tc.Match = td.match
+	}
+
+	providerName := "registry.terraform.io/hashicorp/google"
+	if td.providerName != "" {
+		providerName = td.providerName
+	}
+
+	providerBlock := ""
+	if td.removeProviderBlock != true {
+		providerBlock = fmt.Sprintf(`
+			,"provider_name" : "%s"
+		`, providerName)
+	}
+
+	tc.Object = targettesting.FromJSON(fmt.Sprintf(`
+{
+  "name": "test-name",
+  "type": "test-asset-type",
+  "address": "%s",
+  "change": {}
+	%s
+}
+`, td.address, providerBlock))
+	return tc
+}
+
+var testData = []reviewTestData{
+	{
+		name:      "Basic",
+		address:   "google_compute_global_forwarding_rule.test",
+		wantMatch: true,
+	},
+	{
+		name:      "Basic 2 (wildcard)",
+		match:     match(),
+		address:   "google_compute_global_forwarding_rule.test",
+		wantMatch: true,
+	},
+	{
+		name:      "Only match once.",
+		match:     match(includeAddress("**", "*.*")),
+		address:   "google_compute_global_forwarding_rule.test",
+		wantMatch: true,
+	},
+	{
+		name:      "Match on exact ID",
+		match:     match(includeAddress("google_compute_global_forwarding_rule.test")),
+		address:   "google_compute_global_forwarding_rule.test",
+		wantMatch: true,
+	},
+	{
+		name:      "Does not match address for nested module",
+		match:     match(includeAddress("google_compute_global_forwarding_rule.test")),
+		address:   "module.abc.google_compute_global_forwarding_rule.test",
+		wantMatch: false,
+	},
+	{
+		name:      "name wildcard match",
+		match:     match(includeAddress("google_compute_global_forwarding_rule.*")),
+		address:   "google_compute_global_forwarding_rule.test",
+		wantMatch: true,
+	},
+	{
+		name:      "module wildcard match",
```

Comment:
Can we also add a test case for a root resource (`google_compute_global_forwarding_rule.test`) matching a module wildcard?

Reply:
done